packages/punchcard/lib/appsync/vtl.ts

```diff
@@ -0,0 +1,78 @@
+import { number as numberShape, ShapeGuards, string as stringShape } from '@punchcard/shape';
+import { Shape } from '@punchcard/shape/lib/shape';
+import { VExpression } from './expression';
+import { set, Statement } from './statement';
+import { VFloat, Visitor, VObject, VString } from './vtl-object';
+
+/**
+ * Represents a Velocity Template program.
+ */
+export type VTL<T> = Generator<Statement<VObject | void>, T>;
+
+declare module './vtl-object' {
+  namespace VObject {
+    export function of<T extends Shape>(type: T, expr: VExpression): VObject.Of<T>;
+    export function clone<T extends VObject>(t: T, expr: VExpression): T;
+  }
+}
+VObject.of = function of<T extends Shape>(type: T, expr: VExpression): VObject.Of<T> {
+  return type.visit(Visitor.defaultInstance as any, expr) as any;
+};
+VObject.clone = function clone<T extends VObject>(t: T, expr: VExpression): T {
+  return VObject.of(VObject.typeOf(t), expr) as any;
+};
+
+/**
+ * Type of an ExpressionTemplate factory.
+ *
+ * ```ts
+ * GraphQL.string`${mustBeAGraphQLType}`;
+ * ```
+ */
+export type ExpressionTemplate<T extends Shape> = <Args extends (VObject)[]>(template: TemplateStringsArray,...args: Args) => VTL<VObject.Of<T>>;
+
+/**
+ * Evaluate a VTL template as an instance of some shape.
+ *
+ * The VTL expression must be valid on the RHS of a #set operation.
+ *
+ * ```ts
+ * const str = vtl(string)`hello`;
+ * ```
+ *
+ * Translates to:
+ * ```
+ * #set($var1 = "hello")
+ * ```
+ *
+ * @param type what type to treat the assigned variable
+ */
+export function vtl<T extends Shape>(type: T): ExpressionTemplate<T> {
```

Comment:
Where is this used? Are tagged templates truly a good fit?

Reply:
I think it's an awesome fit. A bit like how styled components lets you write and generate CSS, this lets you write and generate and interact with VTL. I capture the template arguments (which are required to be instances of VObject) and then incorporate them into the VTL template safely.

The `type` argument tells the interpreter what type this data is, so the returned result is an object of that type.

```ts
const sam: VString = ;

const msg = yield* vtl(string)`hello ${sam}`

msg.isEmpty(); // can interact with the msg result like a String in Java
```

Might result in some VTL like:
```
#set($var1 = "sam")
#set($var2 = "hello ${var1}"
```

Note how it generates variables and sets them with values. That's how I perform computations in VTL, each statement sets a variable. I then can use those variables to generate the request/response payloads.

Reply:
Very very cool!